crates/sui-storage/src/event_store/sql.rs

```diff
@@ -214,34 +272,43 @@ const QUERY_BY_MODULE: &str = "SELECT * FROM events WHERE timestamp >= ? AND \
 const QUERY_BY_CHECKPOINT: &str =
     "SELECT * FROM events WHERE checkpoint >= ? AND checkpoint <= ? LIMIT ?";
 
-fn check_limit(limit: usize) -> Result<(), EventStoreError> {
+fn check_limit(limit: usize) -> Result<(), SuiError> {
     if limit <= MAX_LIMIT {
         Ok(())
     } else {
-        Err(EventStoreError::LimitTooHigh(limit))
+        Err(SuiError::TooManyItemsError(MAX_LIMIT as u64))
     }
 }
 
 #[async_trait]
 impl EventStore for SqlEventStore {
-    type EventIt = std::vec::IntoIter<StoredEvent>;
-
     async fn add_events(
         &self,
         events: &[EventEnvelope],
         checkpoint_num: u64,
-    ) -> Result<(), EventStoreError> {
+    ) -> Result<(), SuiError> {
+        let mut cur_seq = self.seq_num.load(Ordering::Acquire);
+        let initial_seq = cur_seq;
+
         // TODO: benchmark
         // TODO: use techniques in https://docs.rs/sqlx-core/0.5.13/sqlx_core/query_builder/struct.QueryBuilder.html#method.push_values
         // to execute all inserts in a single statement?
         // TODO: See https://kerkour.com/high-performance-rust-with-sqlite
         for event in events {
+            // Skip events that have a lower sequence number... which must be same or increasing
```

Comment:
I'm not sure I understand how this can guarantee idempotency. Currently, we process one txn's events in a batch. It's possible that the server crashes in the middle of for loop, leaving us events 50/50 processed. In that case, when we attempt this txn's events after restarts, the 50% of the events will be inserted twice and resulting in duplicates.

Reply:
Hey so this was just a first stab, not meant to be perfect.  I agree that it would not be idempotent if a failure happened in the middle.   There are multiple ways to make this better, I was hoping to do it in subsequent PR and keep iterating.  The simplest method for now would be to make a batch write of all events.

Reply:
How about I add a TODO here?